hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java

```diff
@@ -1434,10 +1437,18 @@ public static boolean recordNeedsRewriteForExtendedAvroTypePromotion(Schema writ
         return needsRewriteToString(writerSchema, true);
       case STRING:
       case BYTES:
+      case FIXED:
+        if (readerSchema.getLogicalType() != null) {
+          // check if logical types are equal
+          return !readerSchema.getLogicalType().equals(writerSchema.getLogicalType());
+        }
         return needsRewriteToString(writerSchema, false);
       case DOUBLE:
         // To maintain precision, you need to convert Float -> String -> Double
-        return writerSchema.getType().equals(Schema.Type.FLOAT);
+        return writerSchema.getType().equals(Schema.Type.FLOAT) && !writerSchema.getType().equals(Schema.Type.STRING);
```

Comment:
how could a type equals `FLOAT` and also `STRING` ?

Reply:
If we can get this [PR](https://github.com/apache/hudi/pull/13654/files) in, I think the `areSchemasProjectionEquivalent` is going to fit the needs of this and has some better testing. I will wait to see if this can be brought into a mergable shape.